### PR TITLE
Add raspian stretch - WIP

### PR DESCRIPTION
**Pull request by stahnma**

### CODE REVIEW

File: configs/components/cpp-hocon.rb [hunk 1]
```diff
@@ -36,16 +41,25 @@
...
   pkg.configure do
     ["#{cmake} \
         #{toolchain} \
+        #{boost_libs} \
         -DCMAKE_VERBOSE_MAKEFILE=ON \
         -DCMAKE_PREFIX_PATH=#{settings[:prefix]} \
         -DCMAKE_INSTALL_PREFIX=#{settings[:prefix]} \
         #{special_flags} \
-        -DBOOST_STATIC=ON \
+        -DBOOST_STATIC=#{boost_static} \
```
Comment (stahnma) [inline configs/components/cpp-hocon.rb]: I use the boost_static logic in several components. I'm wondering if should move that into a project settting?

File: configs/components/cpp-hocon.rb [hunk 2]
```diff
@@ -36,16 +41,25 @@
     end
   end
 
+  if platform.use_native_tools?
+    boost_libs= "-DBOOST_LIBRARYDIR=/usr/lib/arm-linux-gnueabihf/lib -DLEATHERMAN_USE_LOCALES=OFF"
```
Comment (stahnma) [inline configs/components/cpp-hocon.rb]: Look at the variable name here and that I have two things in it. Fix this and then understand LOCALES. GLHF.

File: configs/components/facter.rb [hunk 3]
```diff
@@ -45,13 +45,26 @@
   elsif platform.name =~ /solaris-10/
     pkg.build_requires "http://pl-build-tools.delivery.puppetlabs.net/solaris/10/pl-yaml-cpp-0.5.1.#{platform.architecture}.pkg.gz"
   elsif platform.is_cross_compiled_linux? || platform.name =~ /solaris-11/
-    pkg.build_requires "pl-yaml-cpp-#{platform.architecture}"
+    if platform.use_native_tools?
+      pkg.build_requires "libyaml-cpp-dev:armhf"
```
Comment (stahnma) [inline configs/components/facter.rb]: Change hard-coded armhf to `#{plat.get_architecture}` so this is more reusable.

File: configs/components/facter.rb [hunk 4]
```diff
@@ -165,21 +182,32 @@
 
   unless platform.is_windows?
     special_flags += " -DFACTER_PATH=#{settings[:bindir]} \
-                       -DFACTER_RUBY=#{settings[:libdir]}/$(shell #{ruby} -e 'print RbConfig::CONFIG[\"LIBRUBY_SO\"]') \
+                       -DFACTER_RUBY=/opt/puppetlabs/puppet/lib/libruby.so \
```
Comment (stahnma) [inline configs/components/facter.rb]: That might be worse. Look back into this.

File: configs/components/facter.rb [hunk 5]
```diff
@@ -165,21 +182,32 @@
...
                        -DRUBY_LIB_INSTALL=#{settings[:ruby_vendordir]}"
   end
 
+ if platform.name =~ /debian-9-armhf/
+     boost_libs= "-DBOOST_LIBRARYDIR=/usr/lib/arm-linux-gnueabihf/lib -DLEATHERMAN_USE_LOCALES=OFF"
+     boost_static="OFF"
+     yamlcpp_static = "OFF"
```
Comment (stahnma) [inline configs/components/facter.rb]: Do boost and yaml always have to be the same? (I think so) If so, perhaps use only one variable.

File: configs/components/leatherman.json [hunk 6]
```diff
@@ -1 +1 @@
-{"url":"git://github.com/puppetlabs/leatherman.git","ref":"refs/tags/1.4.0"}
+{"url":"git://github.com/stahnma/leatherman.git","ref":"master"}
```
Comment (stahnma) [inline configs/components/leatherman.json]: Obviously this has to go.

File: configs/components/openssl.rb [hunk 7]
```diff
@@ -47,15 +47,23 @@
...
     ldflags = ''
   elsif platform.is_cross_compiled_linux?
-    pkg.environment "PATH", "/opt/pl-build-tools/bin:$$PATH"
-    pkg.environment "CC", "/opt/pl-build-tools/bin/#{settings[:platform_triple]}-gcc"
 
     cflags = "#{settings[:cflags]} -fPIC"
-    ldflags = "-Wl,-rpath=/opt/pl-build-tools/#{settings[:platform_triple]}/lib -Wl,-rpath=#{settings[:libdir]} -L/opt/pl-build-tools/#{settings[:platform_triple]}/lib"
+    if platform.name =~ /debian-9-armhf/
+      pkg.environment "CC", "/usr/bin/arm-linux-gnueabihf-gcc"
+      ldflags = "-Wl,-rpath=#{settings[:libdir]}"
```
Comment (stahnma) [inline configs/components/openssl.rb]: Could these be pulled from settings?

File: configs/components/openssl.rb [hunk 8]
```diff
@@ -47,15 +47,23 @@
...
+    end
 
-    if platform.architecture == "aarch64"
+    if platform.is_huaweios?
+      ldflags = "-R/opt/pl-build-tools/#{settings[:platform_triple]}/lib -Wl,-rpath=#{settings[:libdir]} -L/opt/pl-build-tools/#{settings[:platform_triple]}/lib"
+      target = 'linux-ppc'
+    elsif platform.architecture == "aarch64"
       target = 'linux-aarch64'
-    elsif platform.name =~ /debian-8-arm/
+    elsif platform.name =~ /debian-[8|9]-arm/
```
Comment (stahnma) [inline configs/components/openssl.rb]: Standardize on either \d or 8|9 but don't use both in the same file.

File: configs/components/openssl.rb [hunk 9]
```diff
@@ -146,7 +154,7 @@
   install_prefix = "INSTALL_PREFIX=/" unless platform.is_windows?
 
   pkg.install do
-    ["#{platform[:make]} #{install_prefix} install"]
+    ["#{platform[:make]} #{install_prefix} install_sw"]
```
Comment (stahnma) [inline configs/components/openssl.rb]: This is an unrelated change and should therefore not be in this PR, but man is it nice.

File: configs/components/pxp-agent.rb [hunk 10]
```diff
@@ -51,18 +55,27 @@
     special_flags += " -DLEATHERMAN_USE_LOCALES=OFF "
   end
 
+  if platform.use_native_tools?
+    boost_libs= "-DBOOST_LIBRARYDIR=/usr/lib/arm-linux-gnueabihf/lib -DLEATHERMAN_USE_LOCALES=OFF"
```
Comment (stahnma) [inline configs/components/pxp-agent.rb]: Use platform triple here or something to abstract the hard-codedness of this string. (and others above)

File: configs/components/ruby-2.4.3.rb [hunk 11]
```diff
@@ -131,7 +131,7 @@
   # Cross-compiles require a hand-built rbconfig from the target system as does Solaris, AIX and Windies
   if platform.is_cross_compiled_linux? || platform.is_solaris? || platform.is_aix? || platform.is_windows?
     pkg.add_source "file://resources/files/ruby_243/rbconfig/rbconfig-#{settings[:platform_triple]}.rb"
-    pkg.build_requires 'runtime' if platform.is_cross_compiled_linux?
+    pkg.build_requires 'runtime' if platform.is_cross_compiled_linux? and ! platform.use_native_tools?
```
Comment (stahnma) [inline configs/components/ruby-2.4.3.rb]: look at `and` vs `&&` and see which is more standard in our repo. Also, this kind of a logic mess.

File: configs/components/ruby-2.4.3.rb [hunk 12]
```diff
@@ -142,7 +142,11 @@
 
 
   if platform.is_deb?
-    pkg.build_requires "zlib1g-dev"
+    if platform.is_cross_compiled?
+      pkg.build_requires "zlib1g-dev:#{platform.architecture}"
```
Comment (stahnma) [inline configs/components/ruby-2.4.3.rb]: Check on this breaking debian-8-armhf, armel or ubuntu for oddball architectures.

File: configs/components/ruby-2.4.3.rb [hunk 13]
```diff
@@ -196,7 +204,7 @@
       special_flags += " --with-baseruby=#{settings[:host_ruby]} --enable-close-fds-by-recvmsg-with-peek "
     end
     pkg.build_requires 'libedit'
-    pkg.build_requires 'runtime'
+    pkg.build_requires 'runtime'
```
Comment (stahnma) [inline configs/components/ruby-2.4.3.rb]: Fix trailing space.

File: configs/components/ruby-2.4.3.rb [hunk 14]
```diff
@@ -295,13 +303,23 @@
...
-      [
-        "#{sed} -i 's|raise|warn|g' #{target_dir}/rbconfig.rb",
-        "mkdir -p #{settings[:datadir]}/doc",
-        "cp #{target_dir}/rbconfig.rb #{settings[:datadir]}/doc",
-        "cp ../rbconfig-#{settings[:platform_triple]}.rb #{target_dir}/rbconfig.rb",
-      ]
+        [
+          "#{sed} -i 's|raise|warn|g' #{target_dir}/rbconfig.rb",
+          "mkdir -p #{settings[:datadir]}/doc",
+          "cp #{target_dir}/rbconfig.rb #{settings[:datadir]}/doc",
```
Comment (stahnma) [inline configs/components/ruby-2.4.3.rb]: Put some comments in here about why you're doing what you're doing.

File: configs/platforms/debian-9-armhf.rb [hunk 15]
```diff
@@ -0,0 +1,25 @@
+platform "debian-9-armhf" do |plat|
+  plat.servicedir "/lib/systemd/system"
+  plat.defaultdir "/etc/default"
+  plat.servicetype "systemd"
+  plat.codename "stretch"
+
+
+  plat.provision_with "export DEBIAN_FRONTEND=noninteractive; apt-get update ; apt-get install -qq -y --no-install-recommends build-essential devscripts make quilt pkg-config debhelper rsync fakeroot gcc-arm-linux-gnueabihf binutils-arm-linux-gnueabihf cpp-arm-linux-gnueabihf g++-arm-linux-gnueabihf curl"
+
+  # You can't use plat.add_build_repository for this because the format is not
```
Comment (stahnma) [inline configs/platforms/debian-9-armhf.rb]: Note that if we had full mirrors of Debian configured in normal ways, we could just run `dpkg --enable-architecture armhf` and it would take care of it all...but alas vmpooler templates not set up like that; and this works regardless.

File: configs/platforms/debian-9-armhf.rb [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+    echo  "deb [arch=#{plat.get_architecture}] http://deb.debian.org/debian #{plat.get_codename} main" >> /etc/apt/sources.list.d/#{plat.get_architecture}.list
+    echo  "deb [arch=#{plat.get_architecture}] http://deb.debian.org/debian #{plat.get_codename}-updates main" >> /etc/apt/sources.list.d/#{plat.get_architecture}.list
+    echo  "deb [arch=#{plat.get_architecture}] http://security.debian.org/debian-security/ #{plat.get_codename}/updates main" >> /etc/apt/sources.list.d/#{plat.get_architecture}.list
+  apt-get -qq update; dpkg --add-architecture #{plat.get_architecture}
+  )
+
+  plat.install_build_dependencies_with "DEBIAN_FRONTEND=noninteractive; apt-get install -qq -qy --no-install-recommends "
+  plat.vmpooler_template "debian-9-x86_64"
+  plat.cross_compiled "true"
+  plat.native_tools "true"
```
Comment (stahnma) [inline configs/platforms/debian-9-armhf.rb]: Remove this once you figure a better way to plumb this logic.

File: configs/projects/puppet-agent.rb [hunk 17]
```diff
@@ -103,16 +103,22 @@
   platform_triple = "powerpc64le-suse-linux" if platform.architecture == "ppc64le" && platform.name =~ /^sles-/
   platform_triple = "powerpc64le-linux-gnu" if platform.architecture == "ppc64el"
   platform_triple = "s390x-linux-gnu" if platform.architecture == "s390x"
-  platform_triple = "arm-linux-gnueabihf" if platform.name == 'debian-8-armhf'
   platform_triple = "arm-linux-gnueabi" if platform.name == 'debian-8-armel'
+  platform_triple = "arm-linux-gnueabihf" if platform.name =~ /debian-[\d]{1}-armhf/
```
Comment (stahnma) [inline configs/projects/puppet-agent.rb]: Yet another regex for this. Arg.

File: configs/projects/puppet-agent.rb [hunk 18]
```diff
@@ -175,9 +181,15 @@
 
   # Define default CFLAGS and LDFLAGS for most platforms, and then
   # tweak or adjust them as needed.
-  proj.setting(:cppflags, "-I#{proj.includedir} -I/opt/pl-build-tools/include")
-  proj.setting(:cflags, "#{proj.cppflags}")
-  proj.setting(:ldflags, "-L#{proj.libdir} -L/opt/pl-build-tools/lib -Wl,-rpath=#{proj.libdir}")
+  if platform.name =~ /debian-9-armhf/
+    proj.setting(:cppflags, "-I#{proj.includedir}")
+    proj.setting(:cflags, "#{proj.cppflags}")
+    proj.setting(:ldflags, "-L/opt/puppetlabs/puppet/lib -L/lib/arm-linux-gnueabihf/ -L/usr/lib/arm-linux-gnueabihf -Wl,-rpath=#{proj.libdir}")
```
Comment (stahnma) [inline configs/projects/puppet-agent.rb]: Change to plat.architecture here for more reusability.

File: configs/components/augeas.rb [hunk 19]
```diff
@@ -1,11 +1,19 @@
 component 'augeas' do |pkg, settings, platform|
   pkg.version '1.10.1'
   pkg.md5sum '6c0b2ea6eec45e8bc374b283aedf27ce'
-  pkg.url "#{settings[:buildsources_url]}/augeas-#{pkg.get_version}.tar.gz"
   pkg.mirror "#{settings[:buildsources_url]}/augeas-#{pkg.get_version}.tar.gz"
+  pkg.url "http://download.augeas.net/augeas-#{pkg.get_version}.tar.gz"
```
Comment (stahnma) [inline configs/components/augeas.rb]: This change is unrelated to this PR and should not be included.

Comment (stahnma) [reply configs/components/augeas.rb]: PR submitted https://github.com/puppetlabs/puppet-agent/pull/1339

File: configs/components/ruby-2.4.3.rb [hunk 20]
```diff
@@ -153,10 +157,14 @@
   end
 
   if platform.is_cross_compiled_linux?
-    pkg.build_requires 'pl-ruby'
+    pkg.build_requires 'pl-ruby' unless platform.use_native_tools?
```
Comment (stahnma) [inline configs/components/ruby-2.4.3.rb]: It's possible that baseruby here isn't needed, but I need to look into it.

File: configs/components/facter.rb [hunk 21]
```diff
@@ -45,13 +45,26 @@
   elsif platform.name =~ /solaris-10/
     pkg.build_requires "http://pl-build-tools.delivery.puppetlabs.net/solaris/10/pl-yaml-cpp-0.5.1.#{platform.architecture}.pkg.gz"
   elsif platform.is_cross_compiled_linux? || platform.name =~ /solaris-11/
-    pkg.build_requires "pl-yaml-cpp-#{platform.architecture}"
+    if platform.use_native_tools?
+      pkg.build_requires "libyaml-cpp-dev:#{platform.architecture}"
+      pkg.requires "libyaml-cpp0.5v5"
+      pkg.requires "libboost-date-time1.62.0"
```
Comment (stahnma) [inline configs/components/facter.rb]: Constrain this to debian.

File: configs/components/leatherman.rb [hunk 22]
```diff
@@ -85,18 +116,29 @@
...
   pkg.configure do
     ["#{cmake} \
         #{toolchain} \
+        #{boost_args} \
         -DLEATHERMAN_GETTEXT=ON \
         -DCMAKE_VERBOSE_MAKEFILE=ON \
         -DCMAKE_PREFIX_PATH=#{settings[:prefix]} \
         -DCMAKE_INSTALL_PREFIX=#{settings[:prefix]} \
         -DCMAKE_INSTALL_RPATH=#{settings[:libdir]} \
+        -DLEATHERMAN_USE_ICU=TRUE \
```
Comment (stahnma) [inline configs/components/leatherman.rb]: Does this have any impact if using pl-boost? It doesn't appear to.

**Comment by mckern** [comment]
> Consider switching amd64/i386 to use the same compile framework

Please don't. Delegate that.

File: configs/components/cpp-hocon.rb [hunk 23]
```diff
@@ -12,8 +12,13 @@
     cmake = "/usr/local/bin/cmake"
     special_flags = "-DCMAKE_CXX_FLAGS='#{settings[:cflags]}'"
   elsif platform.is_cross_compiled_linux?
-    toolchain = "-DCMAKE_TOOLCHAIN_FILE=/opt/pl-build-tools/#{settings[:platform_triple]}/pl-build-toolchain.cmake"
-    cmake = "/opt/pl-build-tools/bin/cmake"
+    if platform.name =~ /debian-9-armhf/
```
Comment (mckern) [inline configs/components/cpp-hocon.rb]: A comment about why this platform is an exception would be helpful for those who come after us.

File: configs/components/dmidecode.rb [hunk 24]
```diff
@@ -1,5 +1,5 @@
 component 'dmidecode' do |pkg, settings, platform|
-  if platform.name == 'el-7-aarch64'
+  if platform.name == 'el-7-aarch64' or platform.name =~ /debian-9-armhf/
```
Comment (mckern) [inline configs/components/dmidecode.rb]: `||`, not `or` please

File: configs/components/facter.rb [hunk 25]
```diff
@@ -45,13 +45,26 @@
...
-    pkg.build_requires "pl-yaml-cpp-#{platform.architecture}"
+    if platform.name =~ /debian-9-armhf/
+      pkg.build_requires "libyaml-cpp-dev:#{platform.architecture}"
+      pkg.requires "libyaml-cpp0.5v5"
+      pkg.requires "libboost-date-time1.62.0"
+      pkg.requires "libboost-thread1.62.0"
+      pkg.requires "libboost-chrono1.62.0"
+      pkg.requires "libboost-atomic1.62.0"
+      pkg.requires "libboost-log1.62.0"
+      pkg.requires "libboost-locale1.62.0"
```
Comment (mckern) [inline configs/components/facter.rb]: Meta commentary: I hate the way Debian drops version numbers into package names. I know this nonsense isn't on you.

Comment (mckern) [reply configs/components/facter.rb]: Relevant commentary: we're sure that these are the only Boost libs needed here?

File: configs/components/leatherman.rb [hunk 26]
```diff
@@ -7,6 +7,32 @@
...
+    pkg.build_requires "libboost-math-dev:#{platform.architecture}"
+    pkg.build_requires "libboost-program-options-dev:#{platform.architecture}"
+    pkg.build_requires "libboost-random-dev:#{platform.architecture}"
+    pkg.build_requires "libboost-serialization-dev:#{platform.architecture}"
+    pkg.build_requires "libboost-signals-dev:#{platform.architecture}"
+    pkg.build_requires "libboost-test-dev:#{platform.architecture}"
+    pkg.build_requires "libboost-system-dev:#{platform.architecture}"
+    pkg.build_requires "libboost-thread-dev:#{platform.architecture}"
+    pkg.build_requires "libboost-timer-dev:#{platform.architecture}"
+    pkg.build_requires "libboost-wave-dev:#{platform.architecture}"
```
Comment (mckern) [inline configs/components/leatherman.rb]: Just require `libboost-all-dev` at this point because this is a hard to follow and big enough that it'll probably drift quickly (compared to other dependencies)

File: configs/components/libxml2.rb [hunk 27]
```diff
@@ -28,12 +28,12 @@
...
   end
 
-  pkg.build_requires 'runtime'
+  pkg.build_requires 'runtime' unless platform.name =~ /debian-9-armhf/
 
   # The system pkg-config has been found to pass incorrect build flags on
   # some (but not all) cross-compiled debian-based platforms:
   if platform.is_cross_compiled? && platform.is_deb?
-    pkg.build_requires "pl-pkg-config" unless platform.name =~ /ubuntu-16\.04-ppc64el/
+    pkg.build_requires "pl-pkg-config" unless platform.name =~ /ubuntu-16\.04-ppc64el/ or platform.name =~ /debian-9-armhf/
```
Comment (mckern) [inline configs/components/libxml2.rb]: `||` instead of `or`

File: configs/components/openssl.rb [hunk 28]
```diff
@@ -14,12 +14,12 @@
     pkg.build_requires 'xorg-x11-util-devel' if platform.name =~ /^sles/
     pkg.build_requires 'xutils-dev' if platform.is_deb?
   elsif platform.is_linux?
-    unless platform.is_fedora? && platform.os_version.delete('f').to_i >= 26
+    unless (platform.is_fedora? && platform.os_version.delete('f').to_i >= 26) or platform !~ /debian-9-armhf/
```
Comment (mckern) [inline configs/components/openssl.rb]: I wanna say `||` instead of `or` but I think once you start putting compound conditionals into `unless`, readability goes out the window. Does this track better for readability?

```ruby
if platform !~ /debian-9-armhf/ || (platform.is_fedora? && platform.os_version.delete('f').to_i >= 26)
  pkg.build_requires 'pl-binutils'
end
```

File: configs/components/openssl.rb [hunk 29]
```diff
@@ -47,15 +47,23 @@
     cflags = settings[:cflags]
     ldflags = ''
   elsif platform.is_cross_compiled_linux?
-    pkg.environment "PATH", "/opt/pl-build-tools/bin:$$PATH"
-    pkg.environment "CC", "/opt/pl-build-tools/bin/#{settings[:platform_triple]}-gcc"
 
     cflags = "#{settings[:cflags]} -fPIC"
-    ldflags = "-Wl,-rpath=/opt/pl-build-tools/#{settings[:platform_triple]}/lib -Wl,-rpath=#{settings[:libdir]} -L/opt/pl-build-tools/#{settings[:platform_triple]}/lib"
+    if platform.name =~ /debian-9-armhf/
+      pkg.environment "CC", "/usr/bin/#{settings[:platform_triple]}-gcc"
```
Comment (mckern) [inline configs/components/openssl.rb]: If you're using the native cross-compile toolchain and modern OpenSSL cross compiling options (`--cross-compile-prefix`), you shouldn't need to specify an explicit `CC`. And The OpenSSL component should *definitely* be updated to use `--cross-compile-prefix` because it'll correctly Do The Thing with OpenSSL internals in ways that the Old Ways may not.

(Of all the components, OpenSSL and Ruby should be refactored the hardest and I understand that you may not have wanted to get stuck in an infinite yak shave. If you choose not to shave this particular yak, could you please add a comment about it and why it was deferred so the next person has some context if they run into edge cases?)

File: configs/components/puppet.json [hunk 30]
```diff
@@ -1 +1 @@
-{"url":"git://github.com/puppetlabs/puppet.git","ref":"ebe7eeffbd1fd8e226198c647891c95a2d9430da"}
+{"url":"git://github.com/puppetlabs/puppet.git","ref":"d70d2b99f54f8db16e24fcfbcdde5b0fee1942f9"}
```
Comment (mckern) [inline configs/components/puppet.json]: Psht, we don't need tags. We have SHAs!

anderson cooper eye roll

File: configs/components/ruby-2.4.3.rb [hunk 31]
```diff
@@ -131,7 +131,7 @@
   # Cross-compiles require a hand-built rbconfig from the target system as does Solaris, AIX and Windies
   if platform.is_cross_compiled_linux? || platform.is_solaris? || platform.is_aix? || platform.is_windows?
     pkg.add_source "file://resources/files/ruby_243/rbconfig/rbconfig-#{settings[:platform_triple]}.rb"
-    pkg.build_requires 'runtime' if platform.is_cross_compiled_linux?
+    pkg.build_requires 'runtime' if platform.is_cross_compiled_linux? && !platform.name =~ /debian-9-armhf/
```
Comment (mckern) [inline configs/components/ruby-2.4.3.rb]: `platform.is_cross_compiled_linux? && platform.name !~ /debian-9-armhf/`

File: configs/components/ruby-2.4.3.rb [hunk 32]
```diff
@@ -153,10 +157,15 @@
   end
 
   if platform.is_cross_compiled_linux?
-    pkg.build_requires 'pl-ruby'
-    special_flags += " --with-baseruby=#{settings[:host_ruby]} "
+    if platform.name =~ /debian-9-armhf/
+      pkg.build_requires "ruby"
+      pkg.environment "CC", "#{settings[:platform_triple]}-gcc"
```
Comment (mckern) [inline configs/components/ruby-2.4.3.rb]: Defining `CC` should not be required if using the native cross-compile support

File: configs/components/ruby-2.4.3.rb [hunk 33]
```diff
@@ -295,13 +304,25 @@
...
-    pkg.install do
-      [
-        "#{sed} -i 's|raise|warn|g' #{target_dir}/rbconfig.rb",
-        "mkdir -p #{settings[:datadir]}/doc",
-        "cp #{target_dir}/rbconfig.rb #{settings[:datadir]}/doc",
-        "cp ../rbconfig-#{settings[:platform_triple]}.rb #{target_dir}/rbconfig.rb",
-      ]
+    if platform.name =~ /debian-9-armhf/
+      # Here we don't overwrite the rbconfig, because the one we create while
+      # building contains exactly what we want (e.g. system compiler, etc)
```
Comment (mckern) [inline configs/components/ruby-2.4.3.rb]: cheering - batman villains

File: configs/components/toolchain.rb [hunk 34]
```diff
@@ -0,0 +1,9 @@
+# This component simply installs a cmake toolchain file for use when building
+# using OS native build toolchains. This is needed before the configure step on
+# leatherman, which is why you can't use a `pkg.install_file` in the leatherman
+# component to make this work.
```
Comment (mckern) [inline configs/components/toolchain.rb]: There's probably a better way to solve this long term but I can live with this.

File: configs/platforms/debian-9-armhf.rb [hunk 35]
```diff
@@ -0,0 +1,24 @@
+platform "debian-9-armhf" do |plat|
+  plat.servicedir "/lib/systemd/system"
+  plat.defaultdir "/etc/default"
+  plat.servicetype "systemd"
+  plat.codename "stretch"
+
+
+  plat.provision_with "export DEBIAN_FRONTEND=noninteractive; apt-get update ; apt-get install -qq -y --no-install-recommends build-essential devscripts make quilt pkg-config debhelper rsync fakeroot gcc-arm-linux-gnueabihf binutils-arm-linux-gnueabihf cpp-arm-linux-gnueabihf g++-arm-linux-gnueabihf curl"
```
Comment (mckern) [inline configs/platforms/debian-9-armhf.rb]: Consider a HEREDOC to split this across a few lines for readability?

File: configs/platforms/debian-9-armhf.rb [hunk 36]
```diff
@@ -0,0 +1,24 @@
...
+
+  # You can't use plat.add_build_repository for this because the format is not
+  # exactly what is expected. This is to enable this specific architecture on
+  # amd64 hosts. (As is the next line)
```
Comment (mckern) [inline configs/platforms/debian-9-armhf.rb]: This might be worth filing a Vanagon bug -- cross-compiles are still not first class compiles in Vanagon and that might be worth working out

File: configs/projects/puppet-agent.rb [hunk 37]
```diff
@@ -103,16 +103,22 @@
   platform_triple = "powerpc64le-suse-linux" if platform.architecture == "ppc64le" && platform.name =~ /^sles-/
   platform_triple = "powerpc64le-linux-gnu" if platform.architecture == "ppc64el"
   platform_triple = "s390x-linux-gnu" if platform.architecture == "s390x"
-  platform_triple = "arm-linux-gnueabihf" if platform.name == 'debian-8-armhf'
   platform_triple = "arm-linux-gnueabi" if platform.name == 'debian-8-armel'
+  platform_triple = "arm-linux-gnueabihf" if platform.name =~ /debian-[\d]-armhf/
```
Comment (mckern) [inline configs/projects/puppet-agent.rb]: I'd also update `gnueabi ` while you're in here:

`platform_triple = "arm-linux-gnueabi" if platform.name =~ /debian-[\d]-armel/`

File: configs/components/facter.rb (hunk 25 above)
Comment (stahnma) [reply configs/components/facter.rb]: I am very sure.

File: configs/components/leatherman.rb (hunk 26 above)
Comment (stahnma) [reply configs/components/leatherman.rb]: Oh, you might think that would work. It doesn't, as some of the boost libs depend on things like python and then you pull in a mulit-arch python which tries to uninstall the existing python and the you basically get into an impossible situation and want to hurt yourself.

File: configs/platforms/debian-9-armhf.rb (hunk 36 above)
Comment (stahnma) [reply configs/platforms/debian-9-armhf.rb]: It's rather archiac setup that would change if we had a proper debian mirror (and we will soon) vs osmirror. Once we switch off it, it won't matter and this can go away.

File: configs/projects/puppet-agent.rb (hunk 37 above)
Comment (stahnma) [reply configs/projects/puppet-agent.rb]: I'm not testing armel (which is eabi without hf). I'd rather not make changes to that right now.

Comment (mckern) [reply configs/projects/puppet-agent.rb]: You're gonna wind up changing it anyway sooner or later

File: configs/components/leatherman.rb (hunk 26 above)
Comment (mckern) [reply configs/components/leatherman.rb]: sad tears

**Comment by mckern** [comment]
You wanna get Branan's eyes on this?

**Comment by stahnma** [comment]
He does. I've it up against mainline PA
